src/Convert/pyconvert.jl

```diff
@@ -307,10 +307,11 @@ end
 
 pyconvert_fix(::Type{T}, func) where {T} = x -> func(T, x)
 
-const PYCONVERT_RULES_CACHE = Dict{Type,Dict{C.PyPtr,Vector{Function}}}()
+const PYCONVERT_RULES_CACHE = Lockable(Dict{Type,Dict{C.PyPtr,Vector{Function}}}())
 
-@generated pyconvert_rules_cache(::Type{T}) where {T} =
-    get!(Dict{C.PyPtr,Vector{Function}}, PYCONVERT_RULES_CACHE, T)
+pyconvert_rules_cache(::Type{T}) where {T} =
+    Base.@lock PYCONVERT_RULES_CACHE _pyconvert_rules_cache!(PYCONVERT_RULES_CACHE[], T)
+@generated _pyconvert_rules_cache!(d::Dict, ::Type{T}) where {T} = get!(Dict{C.PyPtr,Vector{Function}}, d, T)
```

Comment:
I'm noticing this is probably illegal considering world ages (though maybe it's fine so long as nobody defines a new rule?). How important is the caching @cjdoris? Do we need it in a generated function?

Reply:
It has always been illegal to access Dict from a generated function, since that implies mutation, which triggers many different forms of UB in inference all at once